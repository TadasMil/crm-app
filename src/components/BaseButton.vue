<script setup lang="ts" name="BaseButton">
import { ButtonHTMLAttributes, computed } from 'vue'

interface Props {
  type?: ButtonHTMLAttributes['type']
  buttonType?: string
}

const props = withDefaults(defineProps<Props>(), {
  type: 'button',
  buttonType: 'primary',
})

const emit = defineEmits<{
  (e: 'click'): () => void
}>()

const onClick = () => {
  console.log('here')
  emit('click')
}

const typeIsPrimary = computed(() => props.buttonType === 'primary')
const typeIsSuccess = computed(() => props.buttonType === 'success')
</script>

<template>
  <button
    class="rounded-lg px-4 py-1"
    :type="type"
    :class="{
      ' bg-primary font-medium tracking-widest text-white outline-none hover:bg-primary-200':
        typeIsPrimary,
      'border border-success text-success': typeIsSuccess,
    }"
    @click="onClick"
  >
    <slot name="icon" />
    <slot />
  </button>
</template>
