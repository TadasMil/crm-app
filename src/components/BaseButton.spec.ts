import { render, screen } from '@testing-library/vue'
import { describe, expect, it } from 'vitest'
import userEvent from '@testing-library/user-event'

import BaseButton from './BaseButton.vue'

describe('BaseButton.spec.js', () => {
  it('should render primary button', () => {
    const { html } = render(BaseButton, {
      props: {
        buttonType: 'primary',
      },
    })

    expect(html()).toMatchSnapshot()
  })

  it('should render success button', () => {
    const { html } = render(BaseButton, {
      props: {
        buttonType: 'success',
      },
    })

    expect(html()).toMatchSnapshot()
  })

  it('should render default slot', () => {
    const wrapper = render(BaseButton, {
      slots: {
        default: '<h2>default</h2>',
      },
    })

    const buttonDefaultSlotText = screen.findByText('default')

    expect(wrapper).toContain(buttonDefaultSlotText)
  })

  it('should emit click event', async () => {
    const wrapper = render(BaseButton, {
      slots: {
        default: '<h2>default</h2>',
      },
    })

    const user = userEvent.setup()
    wrapper.debug()
    const button = screen.getByRole('button')

    await user.click(button)

    console.log(wrapper.emitted())
    expect(wrapper.emitted()).toHaveBeenCalled()
  })
})
