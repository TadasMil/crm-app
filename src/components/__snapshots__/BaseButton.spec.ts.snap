// Vitest Snapshot v1

exports[`BaseButton.spec.js > should render primary button 1`] = `"<button class=\\"rounded-lg px-4 py-1 bg-primary font-medium tracking-widest text-white outline-none hover:bg-primary-200\\" type=\\"button\\"></button>"`;

exports[`BaseButton.spec.js > should render success button 1`] = `"<button class=\\"rounded-lg px-4 py-1 border border-success text-success\\" type=\\"button\\"></button>"`;
